src/monogame/common/MGM.cpp

```diff
@@ -0,0 +1,183 @@
+// MonoGame - Copyright (C) The MonoGame Team
+// This file is subject to the terms and conditions defined in
+// file 'LICENSE.txt', which is part of this source code package.
+
+#include "csharp_MGM.h"
+
+#include "stl_common.h"
+
+
+struct MGM_Song
+{
+	mgulong duration;
+
+	void (*finishCallback)(void*);
+	void* finishData;
+};
+
+MGM_Song* MGM_Song_Create(const char* mediaFilePath)
+{
+	assert(mediaFilePath != nullptr);
+
+	// TODO: This should detect the media format using
+	// standard native format decoder libraries that are
+	// portable to all our target platforms:
+	//
+	// libvorbis
+	// minimp3
+	// wave ??
+	// mp4 ??
+	// FLAC ??
+	//
+	// It should then spin up thread which decodes the
+	// audio and streams buffers to a native SoundEffect.
+
+	auto song = new MGM_Song();
+	song->duration = 0;
+	song->finishCallback = nullptr;
+	song->finishData = nullptr;
+	return song;
+}
+
+mgulong MGM_Song_GetDuration(MGM_Song* song)
+{
+	assert(song != nullptr);
+	return song->duration;
+}
+
+mgulong MGM_Song_GetPosition(MGM_Song* song)
+{
+	assert(song != nullptr);
+	return 0;
+}
+
+mgfloat MGM_Song_GetVolume(MGM_Song* song)
+{
+	assert(song != nullptr);
+	return 0;
+}
+
+void MGM_Song_SetVolume(MGM_Song* song, mgfloat volume)
+{
+	assert(song != nullptr);
+}
+
+void MGM_Song_Pause(MGM_Song* song)
+{
+	assert(song != nullptr);
+}
+
+void MGM_Song_Play(MGM_Song* song, mgulong startPositionMs, void (*callback)(void*), void* callbackData)
+{
+	assert(song != nullptr);
+
+	song->finishCallback = callback;
+	song->finishData = callbackData;
+}
+
+void MGM_Song_Resume(MGM_Song* song)
+{
+	assert(song != nullptr);
+}
+
+void MGM_Song_Stop(MGM_Song* song)
+{
+	assert(song != nullptr);
+}
+
+void MGM_Song_Destroy(MGM_Song* song)
+{
+	assert(song != nullptr);
+	delete song;
+}
+
+
+struct MGM_Video
+{
+	mguint width;
+	mguint height;
+	mgfloat fps;
+	mgulong duration;
+	mgint cachedFrames;
+};
+
+MGM_Video* MGM_Video_Create(const char* mediaFilePath, mgint cachedFrameNum, mgint& width, mgint& height, mgfloat& fps, mgulong& duration)
+{
+	assert(mediaFilePath != nullptr);
+
+	// TODO: Like Song above we should detect the media
+	// format from a native decoder libraries that are
+	// portable to all our target platforms:
+	//
+	// libtheora - free but limited quality/performance
+	// OpenH264 -
+	// ???
+	//
+	// It should then spin up thread which decodes the
+	// video/audio streams.
+
+	auto video = new MGM_Video();
```

Comment:
Same as the design around `Song`.

In a separate bounty we need to integrate at least Theora.

But ideally we support H264 and a few other video formats.

Reply:
Supporting H264/AAC seems like the best, anything else is bonus and not critical.